File: face_recognition_flutter/my-app/src/services/auth-service.js
// services/auth-service.js
import apiService from './api-service';

class AuthService {
    constructor() {
        this.currentUser = null;
        this.currentToken = null;
        this.tokenKey = 'auth_token';
        this.userKey = 'user_data';
        this.listeners = [];

        // Initialize on creation
        this.initialize();
    }

    // Initialize auth service
    initialize() {
        console.log('AuthService initializing...');
        this.loadUserData();
        console.log('AuthService initialized. Is logged in:', this.isLoggedIn);

        if (this.isLoggedIn) {
            console.log('Current user role:', this.currentUser?.role);
        }
    }

    // Load user data from localStorage
    loadUserData() {
        try {
            const token = localStorage.getItem(this.tokenKey);
            const userJson = localStorage.getItem(this.userKey);

            if (token && userJson) {
                this.currentToken = token;
                this.currentUser = JSON.parse(userJson);
                console.log('User data loaded from localStorage.');
                this.notifyListeners();
            } else {
                console.log('No user data found in localStorage.');
            }
        } catch (error) {
            console.error('Error loading user data from localStorage:', error);
            this.logout();
        }
    }

    // Save login data
    saveLoginData(loginResponse) {
        try {
            this.currentToken = loginResponse.token;
            this.currentUser = loginResponse.user;

            localStorage.setItem(this.tokenKey, loginResponse.token);
            localStorage.setItem(this.userKey, JSON.stringify(loginResponse.user));

            console.log('Login data saved successfully for user:', this.currentUser?.username);
            this.notifyListeners();

            // Update API service token
            if (window.apiService) {
                window.apiService.setToken(loginResponse.token);
            }
        } catch (error) {
            console.error('Failed to save login data:', error);
            throw new Error('Failed to save login data: ' + error.message);
        }
    }

    // Update user data
    updateUser(user) {
        try {
            this.currentUser = user;
            localStorage.setItem(this.userKey, JSON.stringify(user));
            console.log('User data updated successfully for user:', user.username);
            this.notifyListeners();
        } catch (error) {
            console.error('Failed to update user data:', error);
            throw new Error('Failed to update user data: ' + error.message);
        }
    }

    // Logout and clear all data
    logout() {
        try {
            localStorage.removeItem(this.tokenKey);
            localStorage.removeItem(this.userKey);

            this.currentToken = null;
            this.currentUser = null;

            console.log('User logged out. All data cleared.');
            this.notifyListeners();

            // Clear API service token
            if (window.apiService) {
                window.apiService.clearToken();
            }
        } catch (error) {
            console.error('Error during logout:', error);
            // Force clear in memory even if localStorage fails
            this.currentToken = null;
            this.currentUser = null;
            this.notifyListeners();
        }
    }

    // Check permissions based on roles
    hasPermission(allowedRoles) {
        if (!this.currentUser) {
            console.warn('Checking permission but current user is null.');
            return false;
        }

        if (Array.isArray(allowedRoles)) {
            return allowedRoles.includes(this.currentUser.role);
        } else {
            return this.currentUser.role === allowedRoles;
        }
    }

    // Getters
    get isLoggedIn() {
        return this.currentUser !== null && this.currentToken !== null;
    }

    get isStudent() {
        return this.currentUser?.role === 'student';
    }

    get isTeacher() {
        return this.currentUser?.role === 'teacher';
    }

    get isAdmin() {
        return this.currentUser?.role === 'admin';
    }

    get userId() {
        return this.currentUser?.id;
    }

    get userRole() {
        return this.currentUser?.role;
    }

    get userName() {
        return this.currentUser?.full_name || this.currentUser?.fullName;
    }

    get userEmail() {
        return this.currentUser?.email;
    }

    get userUsername() {
        return this.currentUser?.username;
    }

    get token() {
        return this.currentToken;
    }

    get user() {
        return this.currentUser;
    }

    // Auth state listeners
    addListener(callback) {
        this.listeners.push(callback);
        return () => {
            this.listeners = this.listeners.filter(listener => listener !== callback);
        };
    }

    notifyListeners() {
        this.listeners.forEach(callback => {
            try {
                callback({
                    isLoggedIn: this.isLoggedIn,
                    user: this.currentUser,
                    token: this.currentToken
                });
            } catch (error) {
                console.error('Error notifying auth listener:', error);
            }
        });
    }

    // Login method
    async login(credentials) {
        try {
            if (!window.apiService) {
                throw new Error('API Service not available');
            }

            const response = await apiService.login(credentials);

            if (response.success) {
                this.saveLoginData(response.data);
                return response;
            } else {
                throw new Error(response.message || 'Login failed');
            }
        } catch (error) {
            console.error('Login error:', error);
            throw error;
        }
    }

    // Register method
    async register(userData) {
        try {
            if (!window.apiService) {
                throw new Error('API Service not available');
            }

            const response = await window.apiService.register(userData);
            return response;
        } catch (error) {
            console.error('Register error:', error);
            throw error;
        }
    }

    // Get current user profile
    async refreshProfile() {
        try {
            if (!this.isLoggedIn) {
                throw new Error('User not logged in');
            }

            if (!window.apiService) {
                throw new Error('API Service not available');
            }

            const response = await window.apiService.getProfile();

            if (response.success) {
                this.updateUser(response.data);
                return response;
            } else {
                throw new Error(response.message || 'Failed to refresh profile');
            }
        } catch (error) {
            console.error('Refresh profile error:', error);
            // If token is invalid, logout
            if (error.message.includes('401') || error.message.includes('unauthorized')) {
                this.logout();
            }
            throw error;
        }
    }

    // Check if token is expired (basic check)
    isTokenExpired() {
        if (!this.currentToken) return true;

        try {
            // Basic JWT token expiry check
            const payload = JSON.parse(atob(this.currentToken.split('.')[1]));
            const currentTime = Math.floor(Date.now() / 1000);

            return payload.exp < currentTime;
        } catch (error) {
            console.error('Error checking token expiry:', error);
            return true;
        }
    }

    // Auto refresh token if needed
    async checkTokenValidity() {
        if (this.isTokenExpired()) {
            console.log('Token expired, logging out...');
            this.logout();
            return false;
        }
        return true;
    }

    // Change password
    async changePassword(currentPassword, newPassword) {
        try {
            if (!this.isLoggedIn) {
                throw new Error('User not logged in');
            }

            if (!window.apiService) {
                throw new Error('API Service not available');
            }

            const response = await window.apiService.changePassword({
                current_password: currentPassword,
                new_password: newPassword
            });

            return response;
        } catch (error) {
            console.error('Change password error:', error);
            throw error;
        }
    }

    // Update profile
    async updateProfile(profileData) {
        try {
            if (!this.isLoggedIn) {
                throw new Error('User not logged in');
            }

            if (!window.apiService) {
                throw new Error('API Service not available');
            }

            const response = await window.apiService.updateProfile(profileData);

            if (response.success) {
                // Refresh profile to get updated data
                await this.refreshProfile();
                return response;
            } else {
                throw new Error(response.message || 'Failed to update profile');
            }
        } catch (error) {
            console.error('Update profile error:', error);
            throw error;
        }
    }

    // Utility method to get auth headers
    getAuthHeaders() {
        const headers = {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
        };

        if (this.currentToken) {
            headers['Authorization'] = `Bearer ${this.currentToken}`;
        }

        return headers;
    }
}

// Create singleton instance
const authService = new AuthService();

// Export for ES6 import
export default authService;

// Nếu vẫn cần hỗ trợ Node/CommonJS:
if (typeof module !== 'undefined' && module.exports) {
    module.exports = authService;
}

// Gắn vào window nếu chạy browser
if (typeof window !== 'undefined') {
    window.AuthService = AuthService; // class
    window.authService = authService; // instance
}

File: flutter/lib/services/auth_service.dart
import 'dart:convert';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:logger/logger.dart'; // Thêm Logger để debug
import '../models/models.dart'; // Import tất cả models, bao gồm User và LoginResponse

class AuthService {
  // Singleton pattern
  static final AuthService _instance = AuthService._internal();
  factory AuthService() => _instance;
  AuthService._internal();

  static const String _tokenKey = 'auth_token';
  static const String _userKey = 'user_data';

  final FlutterSecureStorage _secureStorage = const FlutterSecureStorage(
    aOptions: AndroidOptions(
      encryptedSharedPreferences: true,
    ),
    iOptions: IOSOptions(
      accessibility: KeychainAccessibility.first_unlock_this_device,
    ),
  );

  final Logger _logger = Logger(); // Khởi tạo Logger

  User? _currentUser;
  String? _currentToken;

  // Getters
  User? get currentUser => _currentUser;
  String? get currentToken => _currentToken;
  bool get isLoggedIn => _currentUser != null && _currentToken != null;

  // Khởi tạo AuthService, nên gọi một lần khi app khởi động
  Future<void> initialize() async {
    _logger.d('AuthService initializing...');
    await _loadUserData();
    _logger.d('AuthService initialized. Is logged in: $isLoggedIn');
    if (isLoggedIn) {
      _logger.d('Current user role: ${_currentUser?.role.name}');
    }
  }

  // Tải dữ liệu người dùng và token từ bộ nhớ an toàn
  Future<void> _loadUserData() async {
    try {
      final token = await _secureStorage.read(key: _tokenKey);
      final userJson = await _secureStorage.read(key: _userKey);

      if (token != null && userJson != null) {
        _currentToken = token;
        _currentUser = User.fromJson(jsonDecode(userJson));
        _logger.d('User data loaded from secure storage.');
      } else {
        _logger.d('No user data found in secure storage.');
      }
    } catch (e) {
      _logger.e('Error loading user data from secure storage: $e');
      // Nếu có lỗi khi đọc, coi như dữ liệu không hợp lệ và xóa đi
      await logout();
    }
  }

  // Lưu token và thông tin người dùng sau khi đăng nhập
  Future<void> saveLoginData(LoginResponse loginResponse) async {
    try {
      _currentToken = loginResponse.token;
      _currentUser = loginResponse.user;

      await _secureStorage.write(key: _tokenKey, value: loginResponse.token);
      await _secureStorage.write(key: _userKey, value: jsonEncode(loginResponse.user.toJson()));
      _logger.i('Login data saved successfully for user: ${_currentUser?.username}');
    } catch (e) {
      _logger.e('Failed to save login data: $e');
      throw Exception('Failed to save login data: $e');
    }
  }

  // Cập nhật thông tin người dùng (ví dụ: sau khi chỉnh sửa hồ sơ)
  Future<void> updateUser(User user) async {
    try {
      _currentUser = user;
      await _secureStorage.write(key: _userKey, value: jsonEncode(user.toJson()));
      _logger.i('User data updated successfully for user: ${user.username}');
    } catch (e) {
      _logger.e('Failed to update user data: $e');
      throw Exception('Failed to update user data: $e');
    }
  }

  // Đăng xuất và xóa tất cả dữ liệu
  Future<void> logout() async {
    try {
      await _secureStorage.delete(key: _tokenKey);
      await _secureStorage.delete(key: _userKey);

      _currentToken = null;
      _currentUser = null;
      _logger.i('User logged out. All data cleared.');
    } catch (e) {
      _logger.e('Error during logout: $e. Forcing clear in memory.');
      // Buộc xóa trong bộ nhớ ngay cả khi có lỗi xóa storage
      _currentToken = null;
      _currentUser = null;
    }
  }

  // Kiểm tra quyền hạn dựa trên vai trò
  bool hasPermission(List<UserRole> allowedRoles) {
    if (_currentUser == null) {
      _logger.w('Checking permission but current user is null.');
      return false;
    }
    return allowedRoles.contains(_currentUser!.role);
  }

  // Getters tiện ích cho vai trò
  bool get isStudent => _currentUser?.role == UserRole.student;
  bool get isTeacher => _currentUser?.role == UserRole.teacher;
  bool get isAdmin => _currentUser?.role == UserRole.admin;

  // Getters cho thông tin người dùng cụ thể từ _currentUser
  int? get userId => _currentUser?.id;
  UserRole? get userRole => _currentUser?.role;
  String? get userName => _currentUser?.fullName;
  String? get userEmail => _currentUser?.email;
  String? get userUsername => _currentUser?.username;
}